### PR TITLE
DateTime timezone fixes

### PR DESCRIPTION
**Pull request by amiraliemami**
Reviewer: @sidravi1 
Estimate: 20mins

---

## Ticket

Fixes: [413](https://idinsight.atlassian.net/browse/AAQ-413) and [536](https://idinsight.atlassian.net/browse/AAQ-536)

### Goal

- Make sure user sees datetimes in their own timezone across the admin app.
- Add loading animation to UD rule addition instead of it saying "Invalid Date"

### Changes

- Updated backend and DB to save datetimes in UTC with timezone information
- Updated backend endpoints to always return datetimes in ISO 8601 format (this automatically allows endpoint to convert dates)
- Updated backend and front-end for Heatmap and Area charts. Backend returns everything in UTC and front-end handles the conversion.
- Added loading animation to UD rule adding

### Future Tasks
- Fix discovered dashboard bugs, see [confluence page](https://idinsight.atlassian.net/wiki/spaces/PD/pages/2627436545/Timezones+fix).
- Refactor dashboard API logic to handle datetimes better? Seems fine though.

## How has this been tested?

Check full app functionality, backend endpoints and frontend full flow and check date and times shown. 

- pytests
- local dev
- docker compose

## Checklist

Fill with `x` for completed. 

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [x] I have updated the automated tests
- [x] I have updated the scripts in `scripts/`

### CODE REVIEW

File: core_backend/tests/api/test_data_api.py [hunk 1]
```diff
@@ -161,9 +162,8 @@ def test_query_data_api_date_filter(
         client: TestClient,
         user1_data: pytest.FixtureRequest,
     ) -> None:
-
-        start_date = datetime.utcnow() - relativedelta(days=start_day)
-        end_date = datetime.utcnow() - relativedelta(days=end_day)
+        start_date = datetime.now(timezone.utc) - relativedelta(days=start_day)
```
Comment (amiraliemami) [inline core_backend/tests/api/test_data_api.py:165]: For some reason this test now fails for one of its cases... Not sure why. Any ideas @sidravi1?

Comment (sidravi1) [reply core_backend/tests/api/test_data_api.py:165]: It is still failing? I see there is an error in the `test_query_data_api_other_user` test though. I'll fix that.

Comment (amiraliemami) [reply core_backend/tests/api/test_data_api.py:171]: yeah just one of these failing for me...

**Review by sidravi1 — APPROVED** [review]
Thank you @amiraliemami!
This has been on my mind for a while - so awesome to see this :)

File: core_backend/migrations/versions/2024_07_22_0e0efc651e4c_update_datetimes_with_utc_timezone.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+depends_on: Union[str, Sequence[str], None] = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.alter_column(
+        "content",
+        "created_datetime_utc",
+        existing_type=postgresql.TIMESTAMP(),
+        type_=sa.DateTime(timezone=True),
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/migrations/versions/2024_07_22_0e0efc651e4c_update_datetimes_with_utc_timezone.py]: Yes it does. I ran the migration and it worked!

File: admin_app/src/app/dashboard/components/Overview.tsx [hunk 3]
```diff
@@ -98,14 +102,30 @@ const Overview: React.FC<OverviewProps> = ({ timePeriod }) => {
     }
   }, [timePeriod, token]);
 
+  const getLocalTime = (time: string) => {
+    // Convert UCT time to local time
+    const date = format(new Date(), "yyyy-MM-dd");
+    const UCTDateTimeString = `${date}T${time}:00.000000Z`;
+    const localDateTimeString = new Date(UCTDateTimeString);
+    const localTimeString = localDateTimeString.toLocaleString(undefined, {
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply admin_app/src/app/dashboard/components/Overview.tsx:110]: I'm ok with parking this for now. It wont make any observable difference at our scale.

**Pull request merged**